### ISSUE TITLE
[FLAKE] Ruby build artifact failure

### ISSUE THREAD

**Issue by lidizheng** [grpc/grpc#22978]
https://fusion.corp.google.com/projectanalysis/history/KOKORO/prod%3Agrpc%2Fcore%2Fmaster%2Flinux%2Fgrpc_build_artifacts

Ruby is experiencing build failure, but I didn't find a direct PR link to it. It might due to external environment changes. This failure is causing CI build `prod:grpc/core/master/linux/grpc_build_artifacts` and `prod:grpc/core/master/linux/grpc_distribtests` to fail over the weekend.

```
2020-05-18 08:04:10,816 ++ dirname tools/run_tests/artifacts/run_in_workspace.sh
+ cd tools/run_tests/artifacts/../../..
++ pwd
+ repo_root=/tmpfs/src/github/grpc
+ export repo_root
+ rm -rf workspace_ruby_native_gem_linux_x64
+ git clone . workspace_ruby_native_gem_linux_x64
Cloning into 'workspace_ruby_native_gem_linux_x64'...
done.
+ git submodule foreach 'cd "${repo_root}/${WORKSPACE_NAME}" \
    && git submodule update --init --reference ${repo_root}/${name} ${name}'
Entering 'third_party/abseil-cpp'
Submodule 'third_party/abseil-cpp' (https://github.com/abseil/abseil-cpp.git) registered for path 'third_party/abseil-cpp'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/abseil-cpp'...
Submodule path 'third_party/abseil-cpp': checked out 'df3ea785d8c30a9503321a3d35ee7d35808f190d'
Entering 'third_party/benchmark'
Submodule 'third_party/benchmark' (https://github.com/google/benchmark) registered for path 'third_party/benchmark'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/benchmark'...
Submodule path 'third_party/benchmark': checked out '090faecb454fbd6e6e17a75ef8146acb037118d4'
Entering 'third_party/bloaty'
Submodule 'third_party/bloaty' (https://github.com/google/bloaty.git) registered for path 'third_party/bloaty'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/bloaty'...
Submodule path 'third_party/bloaty': checked out '73594cde8c9a52a102c4341c244c833aa61b9c06'
Entering 'third_party/boringssl-with-bazel'
Submodule 'third_party/boringssl-with-bazel' (https://github.com/google/boringssl.git) registered for path 'third_party/boringssl-with-bazel'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/boringssl-with-bazel'...
Submodule path 'third_party/boringssl-with-bazel': checked out '1c2769383f027befac5b75b6cedd25daf3bf4dcf'
Entering 'third_party/cares/cares'
Submodule 'third_party/cares/cares' (https://github.com/c-ares/c-ares.git) registered for path 'third_party/cares/cares'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/cares/cares'...
Submodule path 'third_party/cares/cares': checked out 'e982924acee7f7313b4baa4ee5ec000c5e373c30'
Entering 'third_party/envoy-api'
Submodule 'third_party/envoy-api' (https://github.com/envoyproxy/data-plane-api.git) registered for path 'third_party/envoy-api'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/envoy-api'...
Submodule path 'third_party/envoy-api': checked out '8dcc476be69437b505af181a6e8b167fdb101d7e'
Entering 'third_party/gflags'
Submodule 'third_party/gflags' (https://github.com/gflags/gflags.git) registered for path 'third_party/gflags'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/gflags'...
Submodule path 'third_party/gflags': checked out '28f50e0fed19872e0fd50dd23ce2ee8cd759338e'
Entering 'third_party/googleapis'
Submodule 'third_party/googleapis' (https://github.com/googleapis/googleapis.git) registered for path 'third_party/googleapis'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/googleapis'...
Submodule path 'third_party/googleapis': checked out '80ed4d0bbf65d57cc267dfc63bd2584557f11f9b'
Entering 'third_party/googletest'
Submodule 'third_party/googletest' (https://github.com/google/googletest.git) registered for path 'third_party/googletest'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/googletest'...
Submodule path 'third_party/googletest': checked out 'c9ccac7cb7345901884aabf5d1a786cfa6e2f397'
Entering 'third_party/libuv'
Submodule 'third_party/libuv' (https://github.com/libuv/libuv.git) registered for path 'third_party/libuv'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/libuv'...
Submodule path 'third_party/libuv': checked out '15ae750151ac9341e5945eb38f8982d59fb99201'
Entering 'third_party/protobuf'
Submodule 'third_party/protobuf' (https://github.com/google/protobuf.git) registered for path 'third_party/protobuf'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/protobuf'...
Submodule path 'third_party/protobuf': checked out 'd0bfd5221182da1a7cc280f3337b5e41a89539cf'
Entering 'third_party/protoc-gen-validate'
Submodule 'third_party/protoc-gen-validate' (https://github.com/envoyproxy/protoc-gen-validate.git) registered for path 'third_party/protoc-gen-validate'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/protoc-gen-validate'...
Submodule path 'third_party/protoc-gen-validate': checked out '0f2bc6c0fdac9113e3863ea6e30e5b2bd33e3b40'
Entering 'third_party/udpa'
Submodule 'third_party/udpa' (https://github.com/cncf/udpa.git) registered for path 'third_party/udpa'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/udpa'...
Submodule path 'third_party/udpa': checked out 'e8cd3a4bb307e2c810cffff99f93e96e6d7fee85'
Entering 'third_party/zlib'
Submodule 'third_party/zlib' (https://github.com/madler/zlib) registered for path 'third_party/zlib'
Cloning into '/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/third_party/zlib'...
Submodule path 'third_party/zlib': checked out 'cacf7f1d4e3d44d871b605da3b647f07d718623f'
+ echo 'Running in workspace workspace_ruby_native_gem_linux_x64'
Running in workspace workspace_ruby_native_gem_linux_x64
+ cd workspace_ruby_native_gem_linux_x64
+ tools/run_tests/artifacts/build_artifact_ruby.sh
++ uname
++ cut -f 1 -d_
+ SYSTEM=Linux
++ dirname tools/run_tests/artifacts/build_artifact_ruby.sh
+ cd tools/run_tests/artifacts/../../..
+ set +ex
+ '[' Linux == MSYS ']'
+ '[' Linux == MINGW64 ']'
+ '[' Linux == MINGW32 ']'
+ set +ex
Successfully uninstalled rubygems-bundler-1.4.4
Using /usr/local/rvm/gems/ruby-2.4.1
Successfully installed bundler-1.17.3
Parsing documentation for bundler-1.17.3
Installing ri documentation for bundler-1.17.3
Done installing documentation for bundler after 5 seconds
1 gem installed
++ dirname tools/run_tests/helper_scripts/bundle_install_wrapper.sh
+ cd tools/run_tests/helper_scripts/../../..
++ uname
++ cut -f 1 -d_
+ SYSTEM=Linux
+ '[' Linux == Darwin ']'
+ bundle install
Fetching gem metadata from https://rubygems.org/.........
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
google-protobuf-3.12.0-x86_64-linux requires ruby version >= 2.5, which is
incompatible with the current version, ruby 2.4.1p111
+ export DOCKERHUB_ORGANIZATION=grpctesting
+ DOCKERHUB_ORGANIZATION=grpctesting
+ rake gem:native
rake aborted!
LoadError: cannot load such file -- rake/extensiontask
/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/Rakefile:2:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.4.1@global/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
(See full trace by running task with --trace)

2020-05-18 08:04:10,816 FAILED: build_artifact.ruby_native_gem_linux_x64 [ret=1, pid=9112, time=26.9sec]
```

**Comment by apolcyn** [comment]
Looks like the cause of this is that the `google-protobuf` gem just release a 3.12 version on May 16th, https://rubygems.org/gems/google-protobuf/versions, and this release drops support for ruby 2.4 (which this test is running under).

Dropping support for ruby 2.3 and 2.4 seems reasonable though seeing that ruby 2.4 has just been end of life'd (https://www.ruby-lang.org/en/news/2020/04/05/support-of-ruby-2-4-has-ended/).

I think we need to:
a) skip this test for now
b) upgrade all tests currently using ruby 2.3 and  2.4 to use at least ruby 2.5
c) drop support for ruby < 2.5

cc @haberman @dazuma @blowmage

**Comment by jtattermusch** [comment]
@apolcyn  I think dropping support of ruby <2.5 sounds reasonable. No point of supporting something that's past its EOL.

**Comment by apolcyn** [comment]
I chatted with @dazuma and @haberman offline, and we would like to bring back ruby 2.4 support into google-protobuf.

Ruby 2.4 was very recently EOL'd, but there is apparently a lot of client libraries and applications depending on grpc and protobuf that are not ready to migrate off of ruby 2.4, so dropping support at this point may cause problems.

**Comment by lidizheng** [comment]
@apolcyn For deprecated environment, can we ask them to pin to a specific version (or the package manager does it for users)? In that case, we can drop ruby 2.4 for future versions. Only if there is a severe bug, we patch the pinned release.

As for `google-protobuf` change, it might take some time before the change propagate to our CI. Is there a solution to make the BUILD pass sooner?

**Comment by apolcyn** [comment]
> @apolcyn For deprecated environment, can we ask users to pin to a specific version (or gem does it for users)? In that case, we can drop ruby 2.4 for future versions. Only if there is a severe bug, we patch the pinned release.

As noted, there is apparently a lot of client libraries and applications depending on grpc and protobuf that are not ready to migrate off of ruby 2.4. I'd prefer to not release than to do a release that will cause wide breakage. I'll work around this temporarily and hold off on the next ruby release until this is fixed.

I've created https://github.com/grpc/grpc/pull/23015 to work around this, and https://github.com/grpc/grpc/issues/23016 to remove that workaround before next ruby release.

**Comment by apolcyn** [comment]
Protobuf 3.12.1 was just released which adds back support for ruby 2.4 (and 2.3)

I tried installed manually on ruby 2.4 and it worked. So I think this test should actually start passing.
